### PR TITLE
Add cypress to Storybook CI

### PR DESCRIPTION
**Pull request by kroeder**
@ndelangen and I paired on integrating cypress in the Storybook repo. 

#### Run these on the official storybook example
- [x] Routing & navigation
  - [x] Search navigation item
  - [x] Click on nav item
  - [x] Route changes
- [ ] Ejecting the preview
  - [ ] "Url is correct" Knobs + StoryId
- [ ] Keyboard shortcuts (hide/show navigation + hide/show addon panel + fullscreen)
- [ ] Knobs
  - [ ] Go to a story with a knob (maybe with a more advanced setup)
  - [ ] Check if the changes are applied to the preview 

## Tasks
- [x] Integrate cypress so it can run locally
- [x] Check if it works on Windows machines as well
- [ ] Integrate cypress in the Storybook GitHub CI 
- [ ] Add the above mentioned tests

### CODE REVIEW

**Comment by kroeder** [comment]
@ndelangen basic setup is done. I also added typescript 
Can you take a look, please?

File: cypress/helper.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+type StorybookApps = 'official-storybook';
+
+type Addons = 'Knobs';
+
+export const visitExample = (app: StorybookApps, route = '') => {
+  return cy.visit(`${baseUrl}/${app}/${route}`);
+};
+
+export const clickAddon = (addonName: Addons) => {
+  return cy.get(`[role=tablist] button[role=tab]`).contains(addonName);
```
Comment (kroeder) [inline cypress/helper.ts]: Forgot a `click` call

**Comment by ndelangen** [comment]
I'm missing a `cypress run` in the CI somewhere?

**Comment by kroeder** [comment]
I split it into test:e2e-ci and test:e2e-gui
Doesn't t that fit your needs?

**Comment by ndelangen** [comment]
I mean, an actual CI service will have to invoke that then?

I'm not seeing any change to a config of a CI.

**Comment by ndelangen** [comment]
I've added installing and running cypress in the CI & fixed the tests.

I also moved some code into cypress commands. These are unfortunately NOT typescript files. Maybe they can be, if they can be `.ts`, we should migrate them.

**Review by ndelangen — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -113,9 +116,6 @@
     ]
   },
   "browserslist": "defaults",
-  "dependencies": {
```
Comment (shilman) [inline package.json]: ❤️

File: scripts/build-storybooks.js [hunk 3]
```diff
@@ -22,15 +20,19 @@ const logger = console;
...
+    child
+      .on('close', code => {
+        if (code) {
+          reject();
+        } else {
+          resolve();
+        }
+      })
+      .on('error', e => {
+        logger.error(e);
```
Comment (shilman) [inline scripts/build-storybooks.js]: should it reject?

File: cypress/integration/knobs.spec.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { clickAddon, visitExample } from '../helper';
+
+describe('Knobs', () => {
```
Comment (shilman) [inline cypress/integration/knobs.spec.ts]: 😱 😱 😱  awesome!!!

File: package.json [hunk 5]
```diff
@@ -209,14 +211,21 @@
...
     "svelte-jest": "^0.3.0",
     "trash": "^6.0.0",
     "ts-dedent": "^1.1.0",
     "ts-jest": "^24.0.2",
-    "typescript": "^3.4.0"
+    "typescript": "^3.4.0",
+    "wait-on": "^3.3.0"
+  },
+  "optionalDependencies": {
+    "@cypress/webpack-preprocessor": "^4.1.0",
```
[the remark this answers is not in the document]
Comment (kroeder) [reply package.json]: Why are they even optional? For not bothering devs when they bootstrap storybook?

File: scripts/build-storybooks.js (hunk 3 above)
Comment (kroeder) [reply scripts/build-storybooks.js]: Haven't changed the logic of this file
I added `shell: true` (windows fix) and an error output that was missing

It rejects `if (code > 0)` - maybe `code` should be renamed to `exitCode` ?

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (ndelangen) [reply package.json]: in our `yarn bootstrap --install` we do not install optional dependencies.

File: scripts/build-storybooks.js (hunk 3 above)
Comment (ndelangen) [reply scripts/build-storybooks.js]: yeah I should reject I think.

**Comment by ndelangen** [comment]
I added parallelism to the examples generation script, and running them in 4 batches in circleci now.

**Review by shilman — APPROVED** [review]
Looks GREAT! 💯

**Comment by Hypnosphi** [comment]
How do I run it locally? Can you please update CONTRIBUTING.md with instructions?

**Comment by ndelangen** [comment]
https://github.com/storybookjs/storybook/blob/next/CONTRIBUTING.md#2d-run-cypress-tests

Done!

**Pull request merged**